### ISSUE TITLE
Make latest Windows build from Jenkins available to download

### ISSUE THREAD

**Issue by allquixotic** [grpc/grpc#9013]
It seems like a terrific waste of potential *not* to make the Jenkins builds available to the public to download, because on some platforms it is extremely difficult to build gRPC. At least consider making available the very latest successful build, and just overwrite it every time a new build is made?

The reasons why I think this is important are laid out below. 

------------------

I recently had this conversation in IRC:

> [allquixotic] Hi, has anyone been able to recently compile grpc for Windows? I've tried MinGW64 on Fedora 25; MSYS2 on Windows 10; VS2015; with various build systems (cmake -> .sln, cmake -> ninja, just the built-in Makefile) and every time I reliably hit various brick walls. It seems like building for Windows is nigh impossible. Maybe there's a way, but I probably need more explicit instructions than "use MSYS2".
> [ejona] allquixotic: There's constaint builds on Windows in our CI https://grpc-testing.appspot.com/job/gRPC_master_windows/
> [ejona] I'm looking to see what compiler they use.
> [ejona] allquixotic: What language are you trying to use?
> [ejona] Meh. I can't find it quickly. I thought they were on VS2015.
> [allquixotic] ejona: I'm trying to get 64-bit binaries to link into a Qt 5.6.2 DLL and write C++ code (specifically, I need grpc to build https://github.com/googleapis/googleapis )
> [allquixotic] I think WinSxS (Windows side by side DLL compatibility magic) is smart enough that even if the libc used by your grpc build were MinGW, I could still link it into an otherwise VS2015 DLL
> [allquixotic] just need libgrpc.dll, libgrpc++.dll basically
> [allquixotic] was unable to find the binary download from your Jenkins instance though
> [allquixotic] P.S. it'd be great if grpc / boringssl / protobuf wouldn't treat warnings as errors, because as a user I don't _care_ if you have an uninitialized variable, I just want the thing to build :) and it takes a lot of time to figure out for any particular build system / compiler how to go in and find/replace to remove all the flags that turn on warnings as errors
> [ejona] allquixotic: We're not saving the built artifacts in Jenkins, as we won't use them and they add up over time.

I have found the grpc build for the Windows platform to have a ton of "gotchas" and arbitrary failures that I have difficulty breaking through, and have not yet produced a build of grpc for Windows despite spending around **40 hours** on it. I've tried with three versions of Visual Studio, two different installs of MSYS (MSYS "1" and MSYS2), MinGW32 and MinGW64, and I just keep getting stonewalled with different, hard-to-solve problems (and that's not even speaking of the easier to solve problems, like forcibly disabling the `-Werror` / `-WX` flags that cause builds to blow up when it hits a warning...). I even tried using Fedora 25's excellent MinGW64-based cross-compilation stack and scripts like `mingw64-make`, which set up the environment and CFLAGS, etc. perfectly to build 99% of software. Is gRPC really so special that it can't just use a semi-standard build system?

I've been building extremely complex pieces of software such as the Linux kernel, Firefox, Chrome, and Xorg Server since 2003, and if *I'm* having this much difficulty building gRPC, I can only imagine what a relative novice would struggle to do.

Basically, until the Jenkins builds are available to download, it seems only the gRPC maintainers have any ability whatsoever to use gRPC on Windows. That is pretty unfortunate. I'd love to know the secret sauce on your CI server that gets it working.

**Comment by jtattermusch** [comment]
I agree that it is painful to build gRPC C++ on windows, but at the same time it is not true that we are building it for every test run on our testing infrastructure.  We do build C core on windows, which is pretty trivial compared to C++ and can be done in visual studio (and by other means) - so we don't have any secret sauce (btw, all the scripts that run our tests are public in our repository so that's not even possible)

Currently, building C++ on windows is a bit experimental (it works, but we haven't put together all the pieces to make it pleasant and simple to build).
I think the right plan here is to:
1. make C++ easily buildable on windows (the problem is mostly the openssl and protobuf integration).
2. document it properly
3. (publishing artifacts is optional - because everyone can build them as easily. Also it is not clear that the artifacts would be usable by everyone based on the compiler and linker options).

**Comment by jtattermusch** [comment]
cc @nicolasnoble

**Comment by jtattermusch** [comment]
@allquixotic   it seems that after spending some time trying to build C++ on windows, you've resolved at least part of the issues - and many of them might be legitimate issues that we ought to solve. Do you want to come up with a pull request(s) that address some of them?

**Comment by hsaliak** [comment]
Agree with @jtattermusch that the ideal solution is to make this easily buildable on windows and the prioritization. 

> 1. make C++ easily buildable on windows (the problem is mostly the openssl and protobuf integration).
> 1. document it properly
> 1. (publishing artifacts is optional - because everyone can build them as easily. Also it is not clear that the artifacts would be usable by everyone based on the compiler and linker options).

However I think that we should have a long term goal of publishing artifacts in some way that's easily catalogued and would not call it optional. This way binaries can be grabbed if people do not want to build. Not being able to cater to every linker/compiler option is OK, since we would have done (1) and (2).

**Comment by allquixotic** [comment]
@jtattermusch I managed to get grpc++ to build on Windows using some magically good scripts at https://github.com/plasticbox/grpc-windows . This build system wrapper seems to be working fine to make MSVC15 builds for 64-bit, which is exactly what I wanted. The challenge is to figure out how to take this person's hard work and lessons learned and incorporate them into grpc.

**Comment by AndyM84** [comment]
For what it's worth, if you build using `cmake `G "Visual Studio 14 2015" -DgRPC_SSL_PROVIDER="package"` works just fine for me after having cloned the zlib/protobuf sources into the appropriate third_party folders.  Only real notes/problems I ran into were having to downgrade OpenSSL to the latest 1.0.n version because of API/artifact changes in 1.1.n and having to disable building the `grpc++_reflection` build because it doesn't seem to generate the .pb.h/.pb.cc files on its own and even when I generate them myself with `protoc` they are missing the `ServerReflection` type.  Granted, took me quite a while to find all of this via the thrilling practice of reading the CMake files, but it wasn't 40 hours of work either.

Oh, I also had to modify the cmake_install paths for zlib by hand, no matter how many different ways I provided `CMAKE_INSTALL_PREFIX` it just didn't work, mostly because it was hard-coded to the Program Files folder.

**Comment by dzenanz** [comment]
Absence of a currently working CMake build system makes it hard to write cross-platform software based on gRPC.

Linux has a working makefile. Windows build is a nightmare. @plasticbox's version helped. The VSSolution unfortunately uses nuget, requiring manual restore of packages each time my project's solution is regenerated by CMake (which is frequent).

And a lot of ignored pull requests signal it is not really worth investing time into my own modifications.

**Comment by jtattermusch** [comment]
There's been a series of improvements of the CMake build on Windows. Now gRPC should be easy to build and the process is documented and continuously tested.
